### ISSUE TITLE
Upgrading From 20171221 To 20180118 Breakes Configuration

### ISSUE THREAD

**Issue by maxried** [Lochnair/vyatta-wireguard#35]
Platform is ER-X with v1.9.7+hotfix.4
When upgrading wireguard and reloading the device, the wireguard configuration is dropped from the config, but there is a wireguard interface shown in `ip link` that cannot be deleted using the firmware's  tools, but only by `configure`, `save ; exit` and `reboot`.

Trying to configure a wireguard interface on a factory defaulted unit also fails. I could not any config other than the private key to a wireguard interface.

**Comment by Lochnair** [comment]
This is likely caused by the broken modules I pushed. After someone pointed out that the release is broken I uploaded new binaries. So I'm assuming you downloaded the new snapshot before I replaced the assets. If so, simply redownload the latest release, and you should be fine.

You can check if you've got the working module by comparing it's checksum to the one shown below.
```bash
root@s-gate:~# sha256sum /lib/modules/3.10.14-UBNT/kernel/net/wireguard.ko 
1aea58f199bcc65a457a96e2a7811f5de74b87268342082760a4350907873fa4  /lib/modules/3.10.14-UBNT/kernel/net/wireguard.ko
```

**Comment by maxried** [comment]
Actually, I don't think it's the kernel module. The interface comes up just fine. I also just tested a mixture: I installed the 2018 deb, scp'ed the wireguard.ko, reinstalled the 2017 deb and inserted the 2018 wireguard.ko, so that I have the 2017 configuration part and the 2018 kernel module part. This actually works.



I downloaded it earlier today and once again just now:

```
root@riviera:~# sha256sum /lib/modules/3.10.14-UBNT/kernel/net/wireguard.ko
1aea58f199bcc65a457a96e2a7811f5de74b87268342082760a4350907873fa4  /lib/modules/3.10.14-UBNT/kernel/net/wireguard.ko
```

I attached some log files:
[vyatta-commit.log](https://github.com/Lochnair/vyatta-wireguard/files/1668257/vyatta-commit.log)
[vyatta-config-loader.log](https://github.com/Lochnair/vyatta-wireguard/files/1668258/vyatta-config-loader.log)


`/var/log/messages` contains `Jan 26 16:37:32 riviera boot-config-loader: Commit failed at boot.`

**Comment by tssva** [comment]
Can verify that the kernel module works. Installed the latest release and experienced the same issue. Copied over the wg executable from the prior release and the issue was resolved. 

Also if you want to delete the wireguard interface from the failed commit without rebooting you can `sudo ip link del wg0` of course replacing wg0 with whatever your interface name is.

**Comment by Lochnair** [comment]
@maxried Thanks for the logs. I've narrowed down what's going on.

@tssva
This is most likely related to the move from musl to the included glibc in EdgeOS. We originally used musl because wg disliked being run against a 64-bit kernel from 32-bit userspace (due to the use of ioctl). This is no longer needed because of the move to Netlink.

The command that fails now is wg set <if> peer <key> preshared-key /dev/null. Not quite sure what's happening, but I seem to remember someone else having this problem. I'll poke Jason about it.

**Comment by tssva** [comment]
A couple of months ago I pinged Jason because wg set peer preshared-key
/dev/null was not working. It didn't generate an error but it also didn't
remove the preshared-key. He quickly resolved that issue.

**Comment by tssva** [comment]
I have identified the issue in wg and submitted a patch to the WireGuard mailing list. In wg errno is being checked after a fread of the preshared key file. fread doesn't set errno , so it should not be checked. This check causes wg to incorrectly exit with the "fread: Inappropriate ioctl for device" error.

**Comment by Lochnair** [comment]
@tssva
Nice work. I'll build new binaries when Jason has commited the fix to master.

**Comment by Lochnair** [comment]
Should've fixed in 0.0.20180118-2
